lib/manager/gomod/artifacts.ts

```diff
@@ -17,19 +17,88 @@ import type {
   UpdateArtifactsResult,
 } from '../types';
 
-function getPreCommands(): string[] | null {
+function getGoEnvironmentVariables(): NodeJS.ProcessEnv {
+  const goEnvVariables: NodeJS.ProcessEnv = {};
+  const goPrivate: string[] = [];
+  let gitEnvCounter = 0;
+
+  // passthrough the GOPRIVATE environment variable as the first element of the goPrivate array
+  if (process.env.GOPRIVATE) {
+    goPrivate.push(process.env.GOPRIVATE);
+  }
+
+  if (process.env.GIT_CONFIG_COUNT) {
+    // passthrough the GIT_CONFIG_COUNT environment variable as start value of the index count
+    gitEnvCounter = parseInt(process.env.GIT_CONFIG_COUNT, 10);
+    if (Number.isNaN(gitEnvCounter)) {
+      logger.warn(
+        `Found GIT_CONFIG_COUNT env variable, but couldn't parse the value to an integer: ${process.env.GIT_CONFIG_COUNT}. Ignoring it.`
+      );
+      gitEnvCounter = 0;
+    }
+  }
+
   const credentials = find({
     hostType: PLATFORM_TYPE_GITHUB,
     url: 'https://api.github.com/',
   });
-  let preCommands = null;
+
   if (credentials?.token) {
     const token = quote(credentials.token);
-    preCommands = [
-      `git config --global url.\"https://${token}@github.com/\".insteadOf \"https://github.com/\"`, // eslint-disable-line no-useless-escape
-    ];
+    // gitEnvCounter is zero indexed, thus we first create the variables and then increment the counter
+    // prettier-ignore
+    goEnvVariables[`GIT_CONFIG_KEY_${gitEnvCounter}`] = `url.https://${token}@github.com.insteadOf`;
+    // prettier-ignore
+    goEnvVariables[`GIT_CONFIG_VALUE_${gitEnvCounter}`] = `https://github.com`;
+    gitEnvCounter += 1;
+  }
+
+  // get all credentials we have for go using git
+  const goGitCredentials =
+    findAll({
+      hostType: 'go-git',
+    }) || [];
+
+  for (const goGitCredential of goGitCredentials) {
+    // Check that both a token exists and a matchHost
+    if (goGitCredential.token && goGitCredential.matchHost) {
+      // add dummy https protocol if matchHost doesn't include a protocol, so new URL() can parse it
+      if (!/^(?:f|ht)tps?:\/\//.test(goGitCredential.matchHost)) {
+        goGitCredential.matchHost = 'https://' + goGitCredential.matchHost;
+      }
+
+      try {
+        const url: URL = new URL(goGitCredential.matchHost);
+        // For go we use a combined url out of the hostname and pathname
+        // For GitHub that would be something like github.enterprise.com/test-org/test-repo
+        // Allowing to specify different credentials for different paths (organizations/repositories)
+        const goHostUrl = url.hostname + url.pathname;
```

Comment:
Why not using `goGitCredential.matchHost` directly?

Reply:
Mainly because I wanted to get rid of the protocol (e.g. `https://`). Go package names (for e.g. `GORPIVATE`) just consist of the domain followed by an optional path. Therefore `https://` in there would give a wrong value to `GOPRIVATE` and I opted to reconstruct it to allow for more inputs in `matchHost`. 

Alternatively we could as well say that `matchHost` must be without the protocol and only accept it in this case. 
Currently this two `matchHost` would be accepted and effectively the same `github.enterprise.com/test` and `https://github.enterprise.com/test`. Without manually constructing it with `hostname` and `pathname` only the first would work.

Reply:
OK, use the original url for the git config.

I think we should move the git auth config to `lib/util/git/auth.ts` so we can better reuse it for git submodules and git-refs datasource

Reply:
Sounds like a plan, I will do that. 

Should we then (as you mentioned in https://github.com/renovatebot/renovate/pull/10750#discussion_r665981880) just use `git` in the hostRules and require that someone then configures both `git` credentials and manually sets `GOPRIVATE` in their configuration. Background that otherwise if you have `git` credentials for e.g. `github.com` configured and we automatically create `GOPRIVATE` all installs would be from source versus using the GOPROXY.